File: tools/tui/sidecar/probestream_reader.py
"""Python port of ProbeStreamReader from host/ProbeStreamReader.cpp.

Uses OpenOcdTcl for memory access instead of the C++ IProbeMemory interface.
"""

import struct
from dataclasses import dataclass, field
from typing import Callable, Optional

try:
    from .openocd_tcl import OpenOcdTcl
except ImportError:
    from openocd_tcl import OpenOcdTcl

MAGIC = b"ProbeStreamV1\x00\x00\x00"
MAGIC_LEN = 16

OFF_MAGIC = 0
OFF_NUM_UP = 16
OFF_NUM_DOWN = 20
OFF_MAX_UP = 24
OFF_MAX_DOWN = 28
HEADER_SIZE = 32

CH_OFF_PBUFFER = 0
CH_OFF_SIZE = 4
CH_OFF_WROFF = 8
CH_OFF_RDOFF = 12
CH_OFF_FLAGS = 16
CH_DESC_SIZE = 20


def up_channel_offset(index: int) -> int:
    return HEADER_SIZE + index * CH_DESC_SIZE


def down_channel_offset(max_up: int, index: int) -> int:
    return HEADER_SIZE + max_up * CH_DESC_SIZE + index * CH_DESC_SIZE


@dataclass
class ChannelState:
    p_buffer: int = 0
    size: int = 0
    wr_off: int = 0
    rd_off: int = 0
    flags: int = 0
    desc_addr: int = 0


DataCallback = Callable[[int, bytes], None]


class ProbeStreamReader:
    def __init__(self, tcl: OpenOcdTcl, read_mode: str = "auto"):
        self.tcl = tcl
        self.read_mode = read_mode
        self.cb_addr: int = 0
        self.num_up: int = 0
        self.num_down: int = 0
        self.max_up: int = 0
        self.max_down: int = 0
        self.up_channels: list[ChannelState] = []
        self.down_channels: list[ChannelState] = []

    @property
    def attached(self) -> bool:
        return self.cb_addr != 0

    def discover(self, ram_start: int, ram_size: int, scan_chunk_size: int = 1024) -> bool:
        for offset in range(0, ram_size, scan_chunk_size):
            addr = ram_start + offset
            read_len = min(scan_chunk_size, ram_size - offset)
            try:
                data = self.tcl.read_bytes(addr, read_len, self.read_mode)
            except Exception:
                continue
            idx = data.find(MAGIC)
            if idx >= 0:
                self.cb_addr = addr + idx
                return self._read_control_block()
        return False

    def attach(self, cb_addr: int) -> bool:
        try:
            magic = self.tcl.read_bytes(cb_addr, MAGIC_LEN, self.read_mode)
        except Exception:
            return False
        if magic != MAGIC:
            return False
        self.cb_addr = cb_addr
        return self._read_control_block()

    def _read_control_block(self) -> bool:
        self.num_up = self.tcl.read_u32(self.cb_addr + OFF_NUM_UP)
        self.num_down = self.tcl.read_u32(self.cb_addr + OFF_NUM_DOWN)
        self.max_up = self.tcl.read_u32(self.cb_addr + OFF_MAX_UP)
        self.max_down = self.tcl.read_u32(self.cb_addr + OFF_MAX_DOWN)

        if self.num_up == 0 or self.num_up > 64 or self.num_down > 64:
            return False
        if self.max_up < self.num_up or self.max_down < self.num_down:
            return False

        self.up_channels = []
        for i in range(self.num_up):
            ch = ChannelState()
            ch.desc_addr = self.cb_addr + up_channel_offset(i)
            ch.p_buffer = self.tcl.read_u32(ch.desc_addr + CH_OFF_PBUFFER)
            ch.size = self.tcl.read_u32(ch.desc_addr + CH_OFF_SIZE)
            ch.wr_off = 0
            ch.rd_off = 0
            ch.flags = self.tcl.read_u32(ch.desc_addr + CH_OFF_FLAGS)
            self.up_channels.append(ch)

        self.down_channels = []
        for i in range(self.num_down):
            ch = ChannelState()
            ch.desc_addr = self.cb_addr + down_channel_offset(self.max_up, i)
            ch.p_buffer = self.tcl.read_u32(ch.desc_addr + CH_OFF_PBUFFER)
            ch.size = self.tcl.read_u32(ch.desc_addr + CH_OFF_SIZE)
            ch.wr_off = 0
            ch.rd_off = 0
            ch.flags = self.tcl.read_u32(ch.desc_addr + CH_OFF_FLAGS)
            self.down_channels.append(ch)

        return True

    def poll_up(self, cb: Optional[DataCallback] = None) -> int:
        total_read = 0
        for i, ch in enumerate(self.up_channels):
            # One round-trip for both offsets (firmware writes wr, we own rd,
            # so a stale read here is harmless: we just defer leftover bytes
            # to the next poll).
            offsets = self.tcl.mdw_n(ch.desc_addr + CH_OFF_WROFF, 2)
            if len(offsets) < 2:
                continue
            wr_off, rd_off = offsets[0], offsets[1]

            if wr_off == rd_off:
                continue

            if wr_off >= rd_off:
                avail = wr_off - rd_off
            else:
                avail = ch.size - (rd_off - wr_off)

            if avail == 0 or avail >= ch.size:
                continue

            if wr_off > rd_off:
                data = self.tcl.read_bytes(
                    ch.p_buffer + rd_off, avail, self.read_mode
                )
            else:
                part1 = self.tcl.read_bytes(
                    ch.p_buffer + rd_off, ch.size - rd_off, self.read_mode
                )
                part2 = b""
                if wr_off > 0:
                    part2 = self.tcl.read_bytes(
                        ch.p_buffer, wr_off, self.read_mode
                    )
                data = part1 + part2

            # Advance rd_off on target. Single-writer (us) for rd_off, so no
            # lock needed across the read+ack pair.
            self.tcl.write_u32(ch.desc_addr + CH_OFF_RDOFF, wr_off)
            ch.rd_off = wr_off
            ch.wr_off = wr_off

            if data and cb:
                cb(i, data)

            total_read += len(data)

        return total_read

    def write_down(self, channel: int, data: bytes) -> int:
        if channel >= self.num_down or len(data) == 0:
            return 0

        ch = self.down_channels[channel]
        # Single round-trip for both offsets.
        offsets = self.tcl.mdw_n(ch.desc_addr + CH_OFF_WROFF, 2)
        if len(offsets) < 2:
            return 0
        wr_off, rd_off = offsets[0], offsets[1]

        if rd_off > wr_off:
            free = rd_off - wr_off - 1
        else:
            free = (ch.size - 1) - (wr_off - rd_off)

        if free == 0:
            return 0

        to_write = min(len(data), free)

        if wr_off + to_write <= ch.size:
            first_part = min(to_write, ch.size - wr_off)
            self.tcl.write_bytes(ch.p_buffer + wr_off, data[:first_part])
            written = first_part
            new_wr = wr_off + first_part
            if new_wr >= ch.size:
                new_wr = 0
            if written < to_write:
                second_part = to_write - written
                self.tcl.write_bytes(ch.p_buffer, data[written:written + second_part])
                written += second_part
                new_wr = second_part
        else:
            part1 = ch.size - wr_off
            self.tcl.write_bytes(ch.p_buffer + wr_off, data[:part1])
            part2 = to_write - part1
            if part2 > 0:
                self.tcl.write_bytes(ch.p_buffer, data[part1:part1 + part2])
            written = to_write
            new_wr = part2

        self.tcl.write_u32(ch.desc_addr + CH_OFF_WROFF, new_wr)
        ch.wr_off = new_wr
        return written

File: host/ProbeStreamReader.cpp
#include "ProbeStreamReader.h"
#include <algorithm>
#include <cstring>

namespace ProbeStream {

ProbeStreamReader::ProbeStreamReader(IProbeMemory& mem)
    : mem_(mem)
{
}

bool ProbeStreamReader::discover(uint32_t ramStart, uint32_t ramSize, uint32_t scanChunkSize)
{
    std::vector<uint8_t> buf(scanChunkSize + MAGIC_LEN);

    for (uint32_t offset = 0; offset < ramSize; offset += scanChunkSize) {
        uint32_t addr = ramStart + offset;
        uint32_t readLen = std::min(scanChunkSize, ramSize - offset);

        if (!mem_.readMem(addr, buf.data(), readLen))
            continue;

        for (uint32_t i = 0; i + MAGIC_LEN <= readLen; i++) {
            if (std::memcmp(buf.data() + i, MAGIC, MAGIC_LEN) == 0) {
                cbAddr_ = addr + i;
                return readControlBlock();
            }
        }
    }
    return false;
}

bool ProbeStreamReader::attach(uint32_t cbAddr)
{
    uint8_t magic[MAGIC_LEN];
    if (!mem_.readMem(cbAddr, magic, MAGIC_LEN))
        return false;
    if (std::memcmp(magic, MAGIC, MAGIC_LEN) != 0)
        return false;

    cbAddr_ = cbAddr;
    return readControlBlock();
}

bool ProbeStreamReader::readControlBlock()
{
    numUp_   = readU32(cbAddr_ + OFF_NUM_UP);
    numDown_ = readU32(cbAddr_ + OFF_NUM_DOWN);
    maxUp_   = readU32(cbAddr_ + OFF_MAX_UP);
    maxDown_ = readU32(cbAddr_ + OFF_MAX_DOWN);

    if (numUp_ == 0 || numUp_ > 64 || numDown_ > 64)
        return false;
    if (maxUp_ < numUp_ || maxDown_ < numDown_)
        return false;

    upChannels_.resize(numUp_);
    for (uint32_t i = 0; i < numUp_; i++) {
        auto& ch = upChannels_[i];
        ch.descAddr = cbAddr_ + upChannelOffset(i);
        ch.pBuffer = readU32(ch.descAddr + CH_OFF_PBUFFER);
        ch.size    = readU32(ch.descAddr + CH_OFF_SIZE);
        ch.wrOff   = 0;
        ch.rdOff   = 0;
        ch.flags   = readU32(ch.descAddr + CH_OFF_FLAGS);
    }

    downChannels_.resize(numDown_);
    for (uint32_t i = 0; i < numDown_; i++) {
        auto& ch = downChannels_[i];
        ch.descAddr = cbAddr_ + downChannelOffset(maxUp_, i);
        ch.pBuffer = readU32(ch.descAddr + CH_OFF_PBUFFER);
        ch.size    = readU32(ch.descAddr + CH_OFF_SIZE);
        ch.wrOff   = 0;
        ch.rdOff   = 0;
        ch.flags   = readU32(ch.descAddr + CH_OFF_FLAGS);
    }

    return true;
}

bool ProbeStreamReader::refreshChannel(ChannelState& ch)
{
    ch.wrOff = readU32(ch.descAddr + CH_OFF_WROFF);
    ch.rdOff = readU32(ch.descAddr + CH_OFF_RDOFF);
    ch.flags = readU32(ch.descAddr + CH_OFF_FLAGS);
    return true;
}

uint32_t ProbeStreamReader::pollUp(const DataCallback& cb)
{
    uint32_t totalRead = 0;

    for (uint8_t i = 0; i < numUp_; i++) {
        auto& ch = upChannels_[i];
        uint32_t wrOff = readU32(ch.descAddr + CH_OFF_WROFF);
        uint32_t rdOff = readU32(ch.descAddr + CH_OFF_RDOFF);
        ch.flags = readU32(ch.descAddr + CH_OFF_FLAGS);

        if (wrOff == rdOff)
            continue;

        uint32_t avail;
        if (wrOff >= rdOff)
            avail = wrOff - rdOff;
        else
            avail = ch.size - (rdOff - wrOff);

        if (avail == 0 || avail >= ch.size)
            continue;

        std::vector<uint8_t> data(avail);
        uint32_t read = 0;

        if (wrOff > rdOff) {
            mem_.readMem(ch.pBuffer + rdOff, data.data(), avail);
            read = avail;
        } else {
            uint32_t part1 = ch.size - rdOff;
            mem_.readMem(ch.pBuffer + rdOff, data.data(), part1);
            if (wrOff > 0)
                mem_.readMem(ch.pBuffer, data.data() + part1, wrOff);
            read = part1 + wrOff;
        }

        // Advance rdOff on target
        writeU32(ch.descAddr + CH_OFF_RDOFF, wrOff);
        ch.rdOff = wrOff;
        ch.wrOff = wrOff;

        if (read > 0 && cb)
            cb(i, data.data(), read);

        totalRead += read;
    }

    return totalRead;
}

uint32_t ProbeStreamReader::writeDown(uint8_t channel, const uint8_t* data, uint32_t len)
{
    if (channel >= numDown_ || len == 0)
        return 0;

    auto& ch = downChannels_[channel];
    uint32_t wrOff = readU32(ch.descAddr + CH_OFF_WROFF);
    uint32_t rdOff = readU32(ch.descAddr + CH_OFF_RDOFF);

    // Calculate free space
    uint32_t free;
    if (rdOff > wrOff)
        free = rdOff - wrOff - 1;
    else
        free = (ch.size - 1) - (wrOff - rdOff);

    if (free == 0)
        return 0;

    uint32_t toWrite = std::min(len, free);
    uint32_t written = 0;

    if (wrOff + toWrite <= ch.size) {
        // Contiguous write, but check wrap boundary
        uint32_t firstPart = std::min(toWrite, ch.size - wrOff);
        mem_.writeMem(ch.pBuffer + wrOff, data, firstPart);
        written = firstPart;
        uint32_t newWr = wrOff + firstPart;
        if (newWr >= ch.size)
            newWr = 0;

        if (written < toWrite) {
            uint32_t secondPart = toWrite - written;
            mem_.writeMem(ch.pBuffer, data + written, secondPart);
            written += secondPart;
            newWr = secondPart;
        }
        writeU32(ch.descAddr + CH_OFF_WROFF, newWr);
        ch.wrOff = newWr;
    } else {
        uint32_t part1 = ch.size - wrOff;
        mem_.writeMem(ch.pBuffer + wrOff, data, part1);
        uint32_t part2 = toWrite - part1;
        if (part2 > 0)
            mem_.writeMem(ch.pBuffer, data + part1, part2);
        written = toWrite;
        uint32_t newWr = part2;
        writeU32(ch.descAddr + CH_OFF_WROFF, newWr);
        ch.wrOff = newWr;
    }

    return written;
}

uint32_t ProbeStreamReader::readU32(uint32_t addr)
{
    uint8_t buf[4];
    mem_.readMem(addr, buf, 4);
    return buf[0] | (buf[1] << 8) | (buf[2] << 16) | (buf[3] << 24);
}

void ProbeStreamReader::writeU32(uint32_t addr, uint32_t val)
{
    uint8_t buf[4] = {
        static_cast<uint8_t>(val & 0xFF),
        static_cast<uint8_t>((val >> 8) & 0xFF),
        static_cast<uint8_t>((val >> 16) & 0xFF),
        static_cast<uint8_t>((val >> 24) & 0xFF),
    };
    mem_.writeMem(addr, buf, 4);
}

} // namespace ProbeStream
